language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -107,37 +111,23 @@ data CekEnv = CekEnv
 
 makeLenses ''CekEnv
 
--- | The monad the CEK machine runs in.
+-- | The monad the CEK machine runs in. State is inside the ExceptT, so we can
+-- get it back in case of error.
 type CekM = ReaderT CekEnv (ExceptT CekEvaluationException (State ExBudgetState))
 
-spendBoth :: WithMemory Term -> ExCPU -> ExMemory -> CekM ()
-spendBoth term cpu mem = spendCPU term cpu >> spendMemory term mem
-
 spendBudget
-    :: (Ord a, Num a)
-    => CekBudgetMode
-    -> Lens' ExBudgetState a
-    -> a
+    :: ExTallyKey -> WithMemory Term -> ExBudget
     -> CekM ()
-spendBudget Counting    l ex = l += ex
-spendBudget Restricting l ex = do
-    newEx <- l <-= ex
-    when (newEx < 0) $
-        throwingWithCause _EvaluationError (UserEvaluationError CekOutOfExError) Nothing
-
-spendMemory :: WithMemory Term -> ExMemory -> CekM ()
-spendMemory term mem = do
-    modifying exBudgetStateTally
-              (<> (ExTally mempty (ExTallyCounter (singleton (void term) [mem]))))
-    mode <- view cekEnvBudgetMode
-    spendBudget mode (exBudgetStateBudget . exBudgetMemory) mem
-
-spendCPU :: WithMemory Term -> ExCPU -> CekM ()
-spendCPU term cpu = do
+spendBudget key term budget = do
     modifying exBudgetStateTally
-              (<> (ExTally (ExTallyCounter (singleton (void term) [cpu])) mempty))
+              (<> (ExTally (singleton key [(budget, (void term))])))
+    newBudget <- exBudgetStateBudget <%= (\b -> b `plus` (gnegate budget))
     mode <- view cekEnvBudgetMode
-    spendBudget mode (exBudgetStateBudget . exBudgetCPU) cpu
+    case mode of
+        Counting -> pure ()
+        Restricting resb@(ExRestrictingBudget b) ->
+            when ((view exBudgetCPU newBudget) > (view exBudgetCPU b) || (view exBudgetMemory newBudget) > (view exBudgetMemory b)) $ do
```

Comment:
I think this could be a useful function on `ExBudget`. `exceedsBudget :: ExBudget -> ExBudget -> Bool`.